### PR TITLE
General Settings view for Devices

### LINKED ISSUE #584: General Settings view for Devices
**Summary:** 
The console needs to allow users to change general data of a device, similar to how we allow it for applications already.

**Why do we need this?**
It's a basic feature.

**What is already there? What do you see now?**
- SDK support for updating devices: #278 
- The device subview, currently with the overview only #376 
- The "Add Device" view: #573 
  - We should be able to reuse a lot here from the form

**What is missing? What do you want to see?**
- Info on what data can be edited after device creation (cc @htdvisser)
- A way to delete devices
- We don't need wireframes or screen design here

**Environment:**
Console in browser.

**How do you propose to implement this?**
Basically, use the add device form and strip a couple of fields.

**Can you do this yourself and submit a Pull Request?**
Yes.

### PR DESCRIPTION
**Pull request by kschiffer**
#### Summary
Closes #584 

image

#### Changes
- Refactor device data form into a dedicated container
- Add general settings view for devices
- Add option to use `<Tab />`'s as nav elements
- Add submit logic for updating devices
- Add disabled styling to checkboxes and radio buttons
- A couple of minor styling tweaks

#### Notes for Reviewers
I have added the link prop to the `<Tabs />` which will render the tab element as `<NavLink />`, as such also handling active state by itself. I decided to not make it a dedicated component, as the functionality could quite easily be added and it saves us some redundancy. Hope that's ok with you.

#### Release Notes
- Added "General Settings" view for end devices

cc @htdvisser

### CODE REVIEW

File: pkg/webui/locales/xx.json [hunk 1]
```diff
@@ -63,34 +91,7 @@
...
-  "console.views.device-add.messages.supportsClassC": "Xxxxxxxx Xxxxx X",
-  "console.views.device-add.messages.activationMode": "Xxxxxxxxxx Xxxx",
-  "console.views.device-add.messages.otaa": "Xxxx Xxx Xxx Xxxxxxxxxx (XXXX)",
-  "console.views.device-add.messages.abp": "Xxxxxxxxxx Xx Xxxxxxxxxxxxxxx (XXX)",
-  "console.views.device-add.messages.resetWarning": "Xxxxxxxx xx xxxxxxxx xxx xxxxx xxxx xxxxxx xxxxxxxxxxx xxx xxxxxx xxxxxxx",
-  "console.views.device-add.messages.couldNotRetrieveFrequencyPlans": "Xxxxx xxx xxxxxxxx xxx xxxx xx xxxxxxxxx xxxxxxxxx xxxxx",
+  "console.views.device-general-settings.index.updateSuccess": "Xxxxxxxxxxxx xxxxxxx xxx xxxxxx",
   "console.views.device-overview.index.hardware": "Xxxxxxxx",
   "console.views.device-overview.index.brand": "Xxxxx",
   "console.views.device-overview.index.model": "Xxxxx",
```
Comment (bafonins) [inline pkg/webui/locales/xx.json]: please move these to `shared-messages.js`. The gateway model has them as well, see https://github.com/TheThingsNetwork/lorawan-stack/blob/master/api/gateway.proto#L48-L53

File: pkg/webui/console/views/device-general-settings/index.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+export default class DeviceGeneralSettings extends React.Component {
+
+  state = {
+    error: undefined,
+  }
+
+  async handleSubmit (values, { setSubmitting, resetForm }) {
+    const { device } = this.props
+    const { application_id } = device.ids.application_ids
+    const updatedDevice = Object.assign({}, values)
```
Comment (bafonins) [inline pkg/webui/console/views/device-general-settings/index.js]:
```suggestion
    const updatedDevice = {...values}
```

File: pkg/webui/console/views/device-general-settings/index.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+@bind
+export default class DeviceGeneralSettings extends React.Component {
+
+  state = {
+    error: undefined,
+  }
+
+  async handleSubmit (values, { setSubmitting, resetForm }) {
+    const { device } = this.props
+    const { application_id } = device.ids.application_ids
```
Comment (bafonins) [inline pkg/webui/console/views/device-general-settings/index.js]: you can selector here, see https://github.com/TheThingsNetwork/lorawan-stack/blob/master/pkg/webui/lib/selectors/id.js#L17-L20

File: pkg/webui/console/views/device-general-settings/index.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    } else {
+      delete updatedDevice.ids.join_eui
+      delete updatedDevice.ids.dev_eui
+      delete updatedDevice.root_keys
+      delete updatedDevice.resets_join_nonces
+      if (updatedDevice.session.dev_addr) {
+        updatedDevice.ids.dev_addr = updatedDevice.session.dev_addr
+      }
+    }
+    delete updatedDevice.activation_mode
```
Comment (bafonins) [inline pkg/webui/console/views/device-general-settings/index.js]:
```
const { activation_mode, ...updatedDevice } = values
```
and you can remove this + simplifies check on L51

File: pkg/webui/console/views/device-general-settings/index.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+@connect(function ({ device }, props) {
+  return {
+    device: device.device,
+  }
+})
+@bind
+export default class DeviceGeneralSettings extends React.Component {
+
+  state = {
+    error: undefined,
```
Comment (bafonins) [inline pkg/webui/console/views/device-general-settings/index.js]:
```suggestion
    error: '',
```

File: pkg/webui/components/checkbox/checkbox.styl [hunk 6]
```diff
@@ -69,3 +69,7 @@
     -ms-transform: rotate(45deg)
     transform: rotate(45deg)
 
+.disabled
```
Comment (bafonins) [inline pkg/webui/components/checkbox/checkbox.styl]: please add disabled fields to storybook

File: pkg/webui/console/containers/device-data-form/index.js [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+
+  handleOTAASelect () {
+    this.setState({ otaa: true })
+  }
+
+  handleABPSelect () {
+    this.setState({ otaa: false })
+  }
+
+  handleResetsJoinNoncesChange (event) {
```
Comment (bafonins) [inline pkg/webui/console/containers/device-data-form/index.js]: this is not an `event`

File: pkg/webui/console/containers/device-data-form/index.js [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    this.setState({ resets_join_nonces: event })
+  }
+
+  handleResetsFrameCountersChange (event) {
```
Comment (bafonins) [inline pkg/webui/console/containers/device-data-form/index.js]: this is not an `event`

File: pkg/webui/console/containers/device-data-form/index.js [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+        />
+        <Field
+          title={sharedMessages.nwkKey}
+          name="root_keys.nwk_key.key"
+          type="byte"
+          min={16}
+          max={16}
+          placeholder={m.leaveBlankPlaceholder}
+          description={m.nwkKeyDescription}
+
```
Comment (bafonins) [inline pkg/webui/console/containers/device-data-form/index.js]: remove

File: pkg/webui/console/containers/device-data-form/index.js [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+        <SubmitBar>
+          <Button type="submit" message={m.createDevice} />
+        </SubmitBar>
+      </Form>
+    )
+  }
+}
+
+DeviceDataForm.propTypes = {
+  onSubmit: PropTypes.func,
```
Comment (bafonins) [inline pkg/webui/console/containers/device-data-form/index.js]: not required?

File: pkg/webui/console/containers/device-data-form/index.js [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+  error: PropTypes.error,
```
Comment (bafonins) [inline pkg/webui/console/containers/device-data-form/index.js]: add default value

File: pkg/webui/console/containers/device-data-form/index.js [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+  update: PropTypes.bool,
```
Comment (bafonins) [inline pkg/webui/console/containers/device-data-form/index.js]: add default value

File: pkg/webui/console/api/index.js [hunk 13]
```diff
@@ -101,6 +101,7 @@ export default {
     list: ttnClient.Applications.Devices.getAll.bind(ttnClient.Applications.Devices),
     get: ttnClient.Applications.Devices.getById.bind(ttnClient.Applications.Devices),
     create: ttnClient.Applications.Devices.create.bind(ttnClient.Applications.Devices),
+    update: ttnClient.Applications.Devices.updateById.bind(ttnClient.Applications.Devices),
```
Comment (bafonins) [inline pkg/webui/console/api/index.js]: please see how the `applications` and `application` definitions are structured. lets be consistent

File: pkg/webui/lib/selectors/id.js [hunk 14]
```diff
@@ -16,7 +16,7 @@ import getByPath from '../get-by-path'
 
 export const getApplicationId = function (application = {}) {
   return getByPath(application, 'application_id')
-    || getByPath(application, 'application_ids.application_id')
+    || getByPath(application, 'ids.application_id')
```
Comment (bafonins) [inline pkg/webui/lib/selectors/id.js]: this breaks the events component.
my suggestion would be to add an additional case:
```suggestion
    || (getByPath(application, 'application_ids.application_id')
    || (getByPath(application, ''ids.application_id'))
```

File: pkg/webui/console/views/device-general-settings/index.js [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+      }
+    }
+
+    await this.setState({ error: '' })
+    try {
+      const { ids: { device_id }} = this.props.device
+      const changed = diff(device, updatedDevice)
+      await api.device.update(appId, device_id, changed)
+
+      resetForm({ ...values })
```
Comment (bafonins) [inline pkg/webui/console/views/device-general-settings/index.js]:
```suggestion
      resetForm(values)
```
should work as well

File: pkg/webui/lib/selectors/id.js (hunk 14 above)
Comment (bafonins) [reply pkg/webui/lib/selectors/id.js]: This also causes the issue we discussed in slack:
```
checkPropTypes.js:19 Warning: Failed prop type: The prop `emitter` is marked as required in `Event`, but its value is `undefined`.
    in Event (created by DefaultEvent)
    in DefaultEvent (created by List)
    in li (created by ListItem)
    in ListItem (created by List)
    in ol (created by List)
    in div (created by List)
    in List (created by EventsWidget)
    in aside (created by EventsWidget)
    in EventsWidget (created by EventsSubscription)
    in EventsSubscription (created by Connect(EventsSubscription))
```

Comment (kschiffer) [reply pkg/webui/lib/selectors/id.js]: Ahh indeed. Was looking whether it was used elsewhere, but couldn't find it. Will add a case then.

**Pull request merged**